### PR TITLE
Adds the New-and-Improved Training Machine, Training Toolbox!

### PR DESCRIPTION
**Pull request by MrDoomBringer**
_This all started a few weeks ago when I hadn't played for like 6 months lost an honorable toolbox 1v1 against Propser Finely. It was at that point I realized I was sorely rusty, and required a way to rigorously train my clicking abilities. From that, this PR was born._

## About The Pull Request
From what I can tell the target stake in the firing range has literally not worked in years. On top of that, the code for it is kinda all over the place and not very user-friendly, so most people just ignore it. Exactly one map (Icebox station) even has it mapped correctly but still doesn't work there. 

So this PR replaces it with a whole new system:  **The Personnel Training Machine!**
The machine, when activated, will move around wildly in a circle. Has adjustable-but-constrained speed and range.
image
look he even has a little chair to sit people down on

Also, I've added a special toolbox that does no damage, but instead counts the number of hits made against mobs and training machines. Annoy people and practice at the same time!

Both the toolbox and machine will be mapped into dorms. Replacements for both will be available from cargo.

## Why It's Good For The Game
The Toolbox tourney is coming up. What better way to prepare than this?

## Changelog
:cl:
add: Nanotrasen has replaced the broken target stakes with a new model.
add: Training toolboxes have been added! Harmless, but useful for practicing your clicking skills and bugging your crewmembers at the same time.
/:cl:

todo:

- [X] mapping in these things
- [X] cargo packs
- [X] some tgui work
- [X] robot rebellion when emagged
- [X] autodoc

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
As a note, this specific implementation (Notably using the /obj/structure/ type path for the training machine, and using timers instead of SSFastProcess) has received a degree of maintainer approval

**Comment by Tlaltecuhtli** [comment]
make it fight back on emag

**Comment by MrDoomBringer** [comment]
@Tlaltecuhtli that is an excellent idea

File: code/game/objects/items/shooting_range.dm
Comment (Rohesie) [inline code/game/objects/items/shooting_range.dm:13]: Can you only remove the relevant overlays here?

Comment (MrDoomBringer) [reply code/game/objects/items/shooting_range.dm:13]: ok

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:193]: ?

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:176]: Leading new line

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:206]:
```suggestion
	total_hits++
	lap_hits++
```

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:209]: Magic number, move to a define.

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:59]: Early return. Also, better variable names. Preferably, just do `obj/item/target/target` then check `if (!istype(target))`.

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:175]:
```suggestion
	. += "<span class='notice'><b>Click to open control interface.</b></span>"
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by MrDoomBringer** [comment]
@Tlaltecuhtli 
image
he's coming to fuck you up

**Comment by ATH1909** [comment]
add a tracker for how many times you accidentally hit yourself

also, does this move around in a fixed pattern, or does it move around semi-randomly?

**Comment by ATH1909** [comment]
it'd also be nice if it recorded ranged hits, in case you wanna practice your gunslinging with disablers or laser tag guns or something

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:57]:
```suggestion
/obj/structure/training_machine/attackby(obj/item/target/target, mob/user)
```

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:5]:
```suggestion
#define HITS_TO_KILL 9

```

File: code/game/objects/structures/training_machine.dm
Comment (ATH1909) [inline code/game/objects/structures/training_machine.dm:120]: this ignores armor...

File: code/game/objects/structures/training_machine.dm
Comment (ATH1909) [inline code/game/objects/structures/training_machine.dm:167]:
```suggestion
	to_chat(user, "<span class='warning'>You override the training machine's safety protocols, and activate its realistic combat feature. A toolbox pops out of a slot on the top.</span>")
```

File: code/game/objects/structures/training_machine.dm
Comment (ATH1909) [inline code/game/objects/structures/training_machine.dm:166]: is this toolbox dropped when the emagged training machine is destroyed? if so, is that intentional?

Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:166]: maybe and if so yes
each one of these comes with a secret evil toolbox built in don't you know

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:120]: push it to the limit

Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:121]: 9 hits to a drop everyone knows the rules

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:185]: ok fine

**Comment by MrDoomBringer** [comment]
the todos done the reviews addressed now for the map conflicts to roll in

**Comment by MrDoomBringer** [comment]
@ATH1909 the training machine itself technically doesn't count hits done - rather, the cool training toolbox does.

**Comment by MrDoomBringer** [comment]
I could add a hits recieved counter to the training machine but that would involve a little bit of complexity - as we'd have to track hits received a buckled mob - so I'll leave that and other extra features for another PR after this one. For now, I'm just aiming to get this through the door so players can experiment with this before the tourney if they want to

**Review by Twaticus — APPROVED** [review]

File: code/game/objects/structures/training_machine.dm
Comment (Jared-Fogle) [inline code/game/objects/structures/training_machine.dm:145]: Are you sure this is what you want and not `max`? As in, `max(MAX_SPEED - move_speed, 1)`?

File: code/game/objects/items/shooting_range.dm
Comment (Jared-Fogle) [inline code/game/objects/items/shooting_range.dm:71]: ?

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:145]: max works better for this u right

File: code/game/objects/items/shooting_range.dm
Comment (MrDoomBringer) [reply code/game/objects/items/shooting_range.dm:71]: very good mr doom bringer very good

File: code/game/objects/items/shooting_range.dm
Comment (Rohesie) [inline code/game/objects/items/shooting_range.dm:9]: Could you turn this into a lazy list?

File: code/game/objects/items/shooting_range.dm
Comment (Rohesie) [inline code/game/objects/items/shooting_range.dm:14]:
```suggestion
		for (var/bullethole in bullethole_overlays)
```
No need to type check here, right?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:199]: Please put the `Initialize()` and `Destroy()` procs right after the object definition, for better organization.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:35]: Could you clean this variable in `Destroy()` ?

File: tgui/packages/tgui/interfaces/TrainingMachine.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/TrainingMachine.js:33]: why range is stored in local state, when it can be a part of ui_data?

Comment (MrDoomBringer) [reply tgui/packages/tgui/interfaces/TrainingMachine.js]: for future reference, turns out the issue was me forgetting to return TRUE in ui_act()

**Comment by MrDoomBringer** [comment]
This should be ready for further-review/merge. 
Now you can attach a training toolbox to the machine and it'll fight you with it, and thrown training toolboxes counts hits. Emag behavior remains the same.

File: code/game/objects/items/shooting_range.dm
Comment (Rohesie) [inline code/game/objects/items/shooting_range.dm:8]:
```suggestion
	/// Lazylist to keep track of bullet-hole overlays.
```

File: code/game/objects/items/shooting_range.dm
Comment (Rohesie) [inline code/game/objects/items/shooting_range.dm:19]:
```suggestion
		bullethole_overlays = null
```
This is a lazy list, we want it nulled if empty.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:65]:
```suggestion
	. = ..()
	if(.)
		return
```
* Preserve the parent call returns.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:90]: You might want to either handle parent returning a value not to continue the attack or to only call parent when your desired interaction options are unavailable.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:111]:
```suggestion
	vis_contents -= attached_item
```

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:136]: Could you document the new procs?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:170]: I've made use and abuse of timers for my autofire component on tgmc, but then ninjanomnom made a good point in that this overburdens the timer SS, too many things that are not one-time events relying on it.
Could you turn this into a process instead?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:38]:
```suggestion
	var/attack_cooldown = 5 SECONDS
```
* Use time defines.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:40]: Could you use the macros defined in `cooldowns.dm` ?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:185]:
```suggestion
	if (!COOLDOWN_FINISHED(src, attack_cooldown))
```

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:198]:
```suggestion
	COOLDOWN_START(src, attack_cooldown, attack_cooldown_time)
```

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:199]:
```suggestion
	attack_cooldown_time = rand(MIN_ATTACK_DELAY, MAX_ATTACK_DELAY)
```

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:38]:
```suggestion
```
This could be removed, with usage of the cooldown defines, since it won't be var-editable anyway, as it's getting reset on every attack.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:188]:
```suggestion
	for(var/mob/living/carbon/target in view(1, get_turf(src))) //Find adjacent target
```
Do you want to ignore targets which might happen to be on the same tile (teleported, getting there before the mahine became dense, or who knows how else)?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:44]: Can you move this to one of the object damage procs, such as `obj_destruction`? This would produce weird interactions on admin deletes, and explosions on qdeleted objects is not very kosher.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:105]: Could you register to the `attached_item`'s `COMSIG_PARENT_QDELETING` event and handle it properly?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:285]:
```suggestion
	. = ..()
```

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:136]: Sure

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:170]: Is it possible to make the process have a dynamic wait time? The movespeed for this needs to be fairly granular/variable, and after speaking to a maintainer about it they agreed timers might be the best solution here. I might be missing something though

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:188]: Yes, otherwise it will hit mobs buckled to it. I want to train against a monkey buckled to a target stake, after all.
Also, what would be the best way to get a list of mobs adjacent to this, if not oview()?

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:285]: Actually, wouldn't it be better to have
```
if (!..())
  check_hit(hit_atom)
```
so things like getting caught by another mob won't trigger it?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [reply code/game/objects/structures/training_machine.dm:166]: Explained how to handle it via discord.
What maintainer was that?

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [reply code/game/objects/structures/training_machine.dm:184]: Ah, then it's fine.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [reply code/game/objects/structures/training_machine.dm:280]: If you do that then use this pattern:
```cpp
. = ..()
if(!.)
```
Always preserve the parent call's return unless you have a very good reason not to.

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:166]: @Rohesie @SpaceManiac

**Comment by Rohesie** [comment]
You don't generally need to autodoc every proc, just the new ones you add. The redefined ones inherit the documentation at the base.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:146]: Debug info?

**Comment by MrDoomBringer** [comment]
@Rohesie I know, but I felt that documentation could help for some of the overridden procs in any case.

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [inline code/game/objects/structures/training_machine.dm:149]: I'd put these into a proc or macro and call it by both `remove_attached_item()` and this proc.
Because there is behavior you are likely to miss adding to one that would make sense to be on the other, like `handle_density()`

File: code/game/objects/structures/training_machine.dm
Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm:146]: Yes, sorry - didn't push the last commit

File: code/game/objects/structures/training_machine.dm
Comment (Rohesie) [reply code/game/objects/structures/training_machine.dm]: This is still valid, despite the original code is outdated.

Comment (MrDoomBringer) [reply code/game/objects/structures/training_machine.dm]: Went ahead and made remove_attached_item() call this proc

**Comment by MrDoomBringer** [comment]
Just gave another round of testing, no runtimes from what I can tell.

**Pull request merged**
